Automate doxygen frontmatter

joewallwork opened the pull request:
*Don't worry! The vast majority of the lines changed are very minor!*

Since we introduced the pre-commit hook to automatically update date stamps in #660, the date stamps have been updated much more regularly, which is presumably a Good Thing.

However, these trivial changes are becoming real obstacles when rebasing, especially when rebasing over many commits. On several occasions, both @TomMelt and I have found that the vast majority of conflicts that arise are due to differences in when the date stamps were updated. As a result, we feel these frequent date updates are now hampering productivity.

I think we need to reassess how we handle the date stamps. It's not an option to drop them altogether because we require dates and various other pieces of frontmatter in the files to support the Doxygen automatic documentation setup.

In this PR, I propose leaving both the `@date` entry and also the `@file` entry out of all C++ files and provide an automated approach for adding them in when needed. I provide the `scripts/doxygen_prep.sh` script to support these updates, which determines the filename and most recent modification date using `git blame`.

You can run
```sh
for EXTENSION in cpp hpp ipp; do
  find . -name "*.${EXTENSION}" -exec scripts/doxygen_prep.sh {} \;
done
```
from the command line to test it out. Having done so, executing `git diff` should show various changes like
```diff
diff --git a/core/src/CommandLineParser.cpp b/core/src/CommandLineParser.cpp
index ca05fdd9..fa203a8f 100644
--- a/core/src/CommandLineParser.cpp
+++ b/core/src/CommandLineParser.cpp
@@ -1,4 +1,7 @@
 /*!
+ * @file    CommandLineParser.cpp
+ *
+ * @date    05 Jun 2025
  * @author  Tim Spain <timothy.spain@nersc.no>
  * @author  Tim Williams <timothy.williams@nersc.no>
  */
```

Other related changes in this PR:
* Update the Doxygen GitHub Action so that it inserts the `@file` and `@date` information, as appropriate.
* Formatting of frontmatter across all `.cpp`, `.hpp`, and `.ipp` files, for consistency.
* Add `@author` entries to files that were missing them (with the authors chosen determined from who seemed to have contributed most to the `git blame`).
* I propose/clarify in the `README` that `@author` refers to the main author(s) / code owner(s) for the particular file.
* Drop the date stamp update from the `.pre-commit`, as it's no longer relevant.

## Checklist

* [x] Check that the updated Doxgen workflow does indeed work. (It's not triggered by this PR.)

joewallwork committed:
Add Doxygen prep script

joewallwork committed:
Remove @file and @date stamps; fix authors

joewallwork committed:
Add preproc step to Doxygen action

joewallwork committed:
Update README and drop pre-commit no longer needed

joewallwork commented:
Ah, it looks like a few source files got auto-formatted by my editor. Happy to revert those changes, if preferred.

einola reviewed (changes requested):
A very nice change! You should also remove the line mentioning time stamps in the pull request template (in .github/pull_request_template.md)

joewallwork committed:
Drop mention of dates from PR template [skip ci]

joewallwork commented:
> A very nice change! You should also remove the line mentioning time stamps in the pull request template (in .github/pull_request_template.md)

Thanks @einola - done in 85cc28c1.

joewallwork commented:
[Oops, closed by accident :sweat_smile:]

joewallwork committed:
Test Doxygen workflow

joewallwork committed:
Avoid broken pipe issue

joewallwork committed:
set -eux

joewallwork committed:
Filter for doxy_prep script in workflows

joewallwork committed:
Add Doxygen outputs to gitignore

joewallwork committed:
No need to set -x

joewallwork committed:
Set up tmate for Doxygen

joewallwork committed:
Another attempt at restricting workflows [skip ci]

joewallwork committed:
Comment tmate back out

joewallwork committed:
Test Doxygen frontmatter automation (#871)

Testing for #869.

Moving to a separate PR to avoid making noise and running other tests
unnecessarily.

joewallwork commented:
Okay I think I tested this correctly in #871. It's a little hard to do in tmate because I couldn't get the HTML visualisation to work over X11 forwarding, but if I inspect e.g., `docs/html/_para_grid_i_o_8cpp.html` and search for "date" then I see the expected date. Also, it works locally for me FWIW.

Note that I set up the testing such that changes to anything docs related will build the docs but not deploy them to GitHub Pages. That'll only happen upon merging into `main` or `docs`.

joewallwork commented:
Can potentially merge after #877.

joewallwork committed:
Merge branch 'develop' into automate-doxygen-frontmatter

joewallwork committed:
Post-merge fix
